akka-stream/src/main/scala/akka/stream/impl/FoldResourceSinkAsync.scala

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2015-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts.sameThreadExecutionContext
+import akka.stream.ActorAttributes.SupervisionStrategy
+import akka.stream._
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class FoldResourceSinkAsync[T, S](
+  create: () ⇒ Future[S],
+  write:  (S, T) ⇒ Future[Unit],
+  close:  (S) ⇒ Future[Unit]) extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.in")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+
+    val createLogic = new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      private implicit def ec = materializer.executionContext
+
+      private var blockingStream: Option[S] = None
+
+      private var writeFail: Option[Throwable] = None
+
+      private val createdCallback = getAsyncCallback[Try[S]] {
+        case Success(resource) ⇒
+          blockingStream = Some(resource)
+          pull(in)
+
+        case Failure(t) ⇒ failStage(t)
+      }.invokeWithFeedback _
+
+      private val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            writeFail = Some(ex)
+            failStage(ex)
+          case Supervision.Restart ⇒ restartResource()
+          case Supervision.Resume  ⇒ pull(in)
+        }
+      }
+
+      private val pushCallback = getAsyncCallback[Try[Unit]] {
+        case Success(_) ⇒ pull(in)
+        case Failure(t) ⇒ errorHandler(t)
+      }.invoke _
+
+      override def preStart(): Unit = createResource()
+
+      override def onPush(): Unit =
+        blockingStream match {
+          case Some(resource) ⇒
+            try {
+              write(resource, grab(in)).onComplete(pushCallback)(sameThreadExecutionContext)
+            } catch errorHandler
+          case None ⇒
+          // we got a pull but there is no open resource, we are either
+          // currently creating/restarting then the pull will be triggered when creating the
+          // resource completes, or shutting down and then the push does not matter anyway
+        }
+
+      override def postStop(): Unit = {
+        val o = blockingStream
+        blockingStream = None
+        o match {
+          case Some(s) ⇒
+            writeFail.foreach(failStage)
```

Comment:
If `writeFail` is defined I think all paths already have called `failStage` already. Also, there is no `onUpstreamFailure` override, so if upstream fails, this will not be invoked but the matval will be completed with `Done`

Reply:
Ahhh okay that makes sense I missed that one.  I've been banging my head against the wall trying to figure out what I missed. Thanks!